rules/expand_template.bzl

```diff
@@ -0,0 +1,59 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def expand_template_impl(ctx):
+    ctx.actions.expand_template(
+        template = ctx.file.template,
+        output = ctx.outputs.out,
+        substitutions = {
+            k: ctx.expand_location(v, ctx.attr.data)
+            for k, v in ctx.attr.substitutions.items()
+        },
+        is_executable = ctx.attr.is_executable,
+    )
+
+_expand_template = rule(
+    implementation = expand_template_impl,
+    attrs = {
+        "template": attr.label(mandatory = True, allow_single_file = True),
+        "substitutions": attr.string_dict(mandatory = True),
+        "out": attr.output(mandatory = True),
+        "is_executable": attr.bool(default = False, mandatory = False),
+        "data": attr.label_list(allow_files = True),
+    },
+)
+
+def expand_template(name, template, substitutions, out, is_executable = False, data = []):
+    """Template expansion
+    
+    This performs a simple search over the template file for the keys in substitutions,
```

Comment:
Please make it clear that the rule doesn't enforce any special syntax or delimiters on the keys. Something like:

> There is no special syntax for the keys. To avoid conflicts, you would need to explicitly add delimiters to the key strings, for example `"{KEY}"` or `"@KEY@"`.

Reply:
Added comment